### PR TITLE
S 60/el 2439

### PR DESCRIPTION
**Pull request by AlastairMcKee**
Angular 4 Filters migration

### CODE REVIEW

File: docs/app/pages/components/sections/tables/filters/filters.component.ts [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+            library: 'ux-aspects'
+        }]
+    };
+
+    constructor(colorService: ColorService) {
+        this.sparkTrackColor = colorService.getColor('accent').setAlpha(0.2).toRgba();
+        this.sparkBarColor = colorService.getColor('accent').toHex();
+    }
+
+    filtersChanged(event: FilterEvent) {
```
Comment (ashley-hunter) [inline docs/app/pages/components/sections/tables/filters/filters.component.ts]: We should be able to simplify this code quite a bit:

```typescript
constructor(private colorService: ColorService) { }

filtersChanged(event: FilterEvent) {

    // apply a newly added filter
    if (event instanceof FilterAddEvent) {
        this.applyFilter(event.filter);
    }

    // remove an active filter
    if (event instanceof FilterRemoveEvent) {
        this.resetData();
        this.activeFilters = this.activeFilters.filter(filter => filter !== event.filter);
        this.activeFilters.forEach(filter => this.applyFilter(filter));
    }

    // remove all filters
    if (event instanceof FilterRemoveAllEvent) {
        this.resetData();
        this.resetFilters();
    }
}

resetFilters(): void {
    this.activeFilters = [];
}

resetData(): void {
    this.filteredTable = this.table.slice();
}

applyFilter(filter: Filter): void {

    switch (filter.group) {
          
        case 'author':
            this.filteredTable = this.filteredTable.filter(item => item.author === filter.name);
            break;

        case 'status':
            this.filteredTable = this.filteredTable.filter(item => item.active === (filter.name === 'Active'));
            break;
    }

    this.activeFilters.push(filter);
}
```

File: docs/app/pages/components/sections/tables/filters/filters.component.html [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+            <td>This will be triggered when a filter is added, removed or all filters removed. 
+                The event will be an instance of either <code>FilterAddEvent</code>, <code>FilterRemoveEvent</code> 
+                or <code>FilterRemoveAllEvent</code> and will contain the filter being added or removed in a filter property 
+                (remove all will no contain affected filters).</td>
+        </tr>
+
+    </tbody>
+</table>
+
+<h4>Filter Dropdown</h4>
```
Comment (ashley-hunter) [inline docs/app/pages/components/sections/tables/filters/filters.component.html]: This is an h4 whereas filter dynamic is an h3. Should probably make consistent

File: docs/app/data/components-page.json [hunk 3]
```diff
@@ -834,13 +834,20 @@
                 },
                 {
                     "title": "Filters",
+                    "component": "ComponentsFiltersComponent",
+                    "version": "Angular"
```
Comment (ashley-hunter) [inline docs/app/data/components-page.json]: Add a usage section

File: docs/app/pages/components/sections/tables/filters/filters.component.html [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+                <div class="hpe-icon" [class.text-accent]="item.active" [class.hpe-checkmark]="item.active"></div>
+            </td>
+        </tr>
+    </tbody>
+</table>
+
+<hr>
+
+<p>Filters can be implemented to offer users a way to filter a set of data, allowing them to apply and remove conditions.
+    Filters are contianed within a Filters Container compoenent which contains a clear all button for removing all active filters at once</p>
```
Comment (ashley-hunter) [inline docs/app/pages/components/sections/tables/filters/filters.component.html]: Fix typos: `contianed` within a Filters Container `compoenent`

File: docs/app/pages/components/sections/tables/filters/filters.component.html [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+            <td><b>Filter</b></td>
+            <td>Sets the initial filter state, this should be set to the state where nothing has been filtered as it will revert to this when the filters have been removed.</td>
+        </tr>
+
+    </tbody>
+</table>
+
+<h4>Filter Options</h4>
+
+<p>A filter object can contain the follwing properties:</p>
```
Comment (ashley-hunter) [inline docs/app/pages/components/sections/tables/filters/filters.component.html]: Fix typo `follwing`

File: docs/app/pages/components/sections/tables/filters/filters.component.html [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+
+<p>When there are a large number of filter options, the Dynamic Filter can be used to replace the regular 
+    dropdown with a text search, supporting dropdown typeahead. This is configurable to fallback to the 
+    standard implementation when the list of filters falls beneath a certain length. An example of this is shown 
+    as Author in the demo above. 
+</p>
+
+<p>The dynamic filter can be added inside the filter container using the <code>ux-filter-dynamic</code> component.</p>
+
+<p>The following attributes fcan be used to customise the behavior of the component:</p>
```
Comment (ashley-hunter) [inline docs/app/pages/components/sections/tables/filters/filters.component.html]: Fix typo: `fcan`

Comment (ashley-hunter) [inline docs/app/pages/components/sections/tables/filters/filters.component.html]: Would be good to add a section here about Custom Filters and how teams can create their own filters - similar to the custom facets section.

**Comment by AlastairMcKee** [comment]
Made changes and added a new section for custom component with a demo and plunker

File: src/components/filters/filter-dropdown/filter-dropdown.component.html [hunk 7]
```diff
@@ -0,0 +1,13 @@
+<div class="btn-group" dropdown>
+    <button dropdownToggle type="button" class="filter-dropdown btn dropdown-toggle">{{ selected?.title }} 
+        <span class="caret"></span>
+    </button>
+    <ul *dropdownMenu class="dropdown-menu" role="menu">
+        <li class="dropdown-list-item" *ngFor="let filter of filters" role="menuitem">
+            <a class="dropdown-item" (click)="selectFilter(filter)">
```
Comment (ashley-hunter) [inline src/components/filters/filter-dropdown/filter-dropdown.component.html]: image

Text text in the dropdown should probably all be aligned the same regardless of whether the checkmark is there or not

File: docs/app/pages/components/sections/tables/custom-filters-component/custom-filters.component.ts [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+        ],
+        modules: [{
+            imports: ['FiltersModule', 'RadioButtonModule'],
+            library: 'ux-aspects'
+        }, {
+            imports: ['SampleFilterCustomComponent'],
+            library: './sample-filter.component',
+            declaration: true
+        }, {
+            imports: ['BsDropdownModule', 'DatepickerModule'],
```
Comment (ashley-hunter) [inline docs/app/pages/components/sections/tables/custom-filters-component/custom-filters.component.ts]: Is datepickermodule required here?

File: docs/app/services/plunker/plunker.service.ts [hunk 9]
```diff
@@ -49,7 +48,21 @@ export class PlunkerService {
 
             // create list of module imports
             plunk.modules.filter(module => !module.declaration).forEach(module => {
-                const moduleImportProviders = module.providers || module.imports;
+
+                let moduleImports: any;
+                if (module.forRoot) {
```
Comment (ashley-hunter) [inline docs/app/services/plunker/plunker.service.ts]: Should be able to simplify this to:
```typescript
let moduleImports: string | string[];

if (module.forRoot) {
    moduleImports = Array.isArray(module.imports) ? module.imports.map(imp => `${imp}.forRoot()`) : `${module.imports}.forRoot()`;
}

const moduleImportProviders = module.providers || moduleImports || module.imports;
```

**Comment by AlastairMcKee** [comment]
Made requested changes

**Pull request merged**